src/detect-flow-pkts.c

```diff
@@ -23,114 +23,141 @@
+    if (df == NULL)
+        return -1;
 
-void DetectFlowPktsToClientRegister(void)
-{
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].name = "flow.pkts_toclient";
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].desc = "match flow number of packets to client";
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].url = "/rules/flow-keywords.html#flow-pkts_toclient";
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].Match = DetectFlowPktsToClientMatch;
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].Setup = DetectFlowPktsToClientSetup;
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].Free = DetectFlowPktsToClientFree;
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].SupportsPrefilter =
-            PrefilterFlowPktsToClientIsPrefilterable;
-    sigmatch_table[DETECT_FLOW_PKTS_TO_CLIENT].SetupPrefilter = PrefilterSetupFlowPktsToClient;
-}
+    df->pkt_data = du32;
+    df->dir = DETECT_FLOW_TO_CLIENT;
 
-static int DetectFlowPktsToServerMatch(
-        DetectEngineThreadCtx *det_ctx, Packet *p, const Signature *s, const SigMatchCtx *ctx)
-{
-    if (p->flow == NULL) {
-        return 0;
+    if (SigMatchAppendSMToList(
+                de_ctx, s, DETECT_FLOW_PKTS, (SigMatchCtx *)df, DETECT_SM_LIST_MATCH) == NULL) {
+        DetectFlowPktsFree(de_ctx, df);
+        return -1;
     }
-    uint32_t nb = p->flow->todstpktcnt;
+    s->flags |= SIG_FLAG_REQUIRE_PACKET;
 
-    const DetectU32Data *du32 = (const DetectU32Data *)ctx;
-    return DetectU32Match(nb, du32);
+    return 0;
 }
 
-static void DetectFlowPktsToServerFree(DetectEngineCtx *de_ctx, void *ptr)
+static int DetectFlowPktsSetup(DetectEngineCtx *de_ctx, Signature *s, const char *rawstr)
 {
-    rs_detect_u32_free(ptr);
-}
-
-static int DetectFlowPktsToServerSetup(DetectEngineCtx *de_ctx, Signature *s, const char *rawstr)
-{
-    DetectU32Data *du32 = DetectU32Parse(rawstr);
-    if (du32 == NULL)
-        return -1;
-
-    if (SigMatchAppendSMToList(de_ctx, s, DETECT_FLOW_PKTS_TO_SERVER, (SigMatchCtx *)du32,
-                DETECT_SM_LIST_MATCH) == NULL) {
-        DetectFlowPktsToServerFree(de_ctx, du32);
+    DetectFlow *df = SCCalloc(1, sizeof(DetectFlow));
+    if (df == NULL)
         return -1;
+    char copy[strlen(rawstr) + 1];
+    strlcpy(copy, rawstr, sizeof(copy));
+    char *context = NULL;
+    char *token = strtok_r(copy, ",", &context);
+    bool dir_set = false;
+    uint8_t num_tokens = 1;
+    while (token != NULL) {
+        if (num_tokens > 2)
+            goto error;
+
+        while (*token != '\0' && isblank(*token)) {
+            token++;
+        }
+        if (strlen(token) == 0) {
+            goto next;
+        }
+
+        num_tokens++;
+
+        if (strcmp(token, "toserver") == 0) {
+            df->dir = DETECT_FLOW_TO_SERVER;
+        } else if (strcmp(token, "toclient") == 0) {
+            df->dir = DETECT_FLOW_TO_CLIENT;
+        }
+
+        if (dir_set) {
+            DetectU32Data *du32 = DetectU32Parse(token);
```

Comment:
feel it would be cleaner to separate the strtok_r loop from acting on the results of the loop